### PR TITLE
Optimizations of union, intersection and iterators of concise bitsets (part of #3798)

### PR DESCRIPTION
**Pull request by leventov**
- Port of metamx/extendedset#10, metamx/extendedset#13, metamx/extendedset#14, metamx/extendedset#15, metamx/bytebuffer-collections@9b199e334990d16defe003e2b436240c9f60a156, metamx/bytebuffer-collections#38 to Druid, i. e. the changes in metamx/extendedset and metamx/bytebuffer-collections since bytebuffer-collections ([PR](https://github.com/druid-io/druid/pull/3647)) and extendedset ([PR](https://github.com/druid-io/druid/pull/3694)) were moved to druid-io/druid
 - Removed unused code in the extendedset module

Benchmark results could be found in the description of metamx/extendedset#13. Generally after optimizations union and intersection of concise bitmaps are 2-3 times faster.

Some of those changes are already reviewed by @egor-ryashin, @gianm and @logarithm.

Druid binaries which include this patch are running in our production for more than a month now.

Part of #3798.

### CODE REVIEW

File: extendedset/src/main/java/io/druid/extendedset/intset/ConciseSet.java [hunk 1]
```diff
@@ -75,7 +73,7 @@
    * User for <i>fail-fast</i> iterator. It counts the number of operations
    * that <i>do</i> modify {@link #words}
    */
-  protected transient volatile int modCount = 0;
+  protected transient int modCount = 0;
```
Comment (gianm) [inline extendedset/src/main/java/io/druid/extendedset/intset/ConciseSet.java]: Looks like this is never being read, even in subclasses. We should just get rid of it if it's not doing anything useful.

Comment (leventov) [inline extendedset/src/main/java/io/druid/extendedset/intset/ConciseSet.java]: Removed

File: extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java [hunk 2]
```diff
@@ -230,43 +288,46 @@ private static void addAndCompact(IntList set, int wordToAdd)
...
-      int position = Integer.numberOfTrailingZeros(last) + 1;
-      if (ConciseSetUtils.isAllZerosLiteral(wordToAdd)) {
-        newWord = 0x00000001 | (position << 25);
-      } else if (ConciseSetUtils.isZeroSequence(wordToAdd) && ConciseSetUtils.getFlippedBit(wordToAdd) == -1) {
-        newWord = (wordToAdd + 1) | (position << 25);
-      }
-    } else if (ConciseSetUtils.isLiteralWithSingleZeroBit(last)) {
-      int position = Integer.numberOfTrailingZeros(~last) + 1;
+    } else {
+      // last is "one sequence"
```
Comment (gianm) [inline extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java]: `assert ConciseSetUtils.isOneSequence(last)`?

**Comment by leventov** [comment]
@gianm thanks for review. Replaced comments with assertions.

File: extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java [hunk 3]
```diff
@@ -550,37 +534,83 @@ public int compare(WordHolder h1, WordHolder h2)
...
+          WordIterator i = theQ[qIndex];
+          int w = i.getWord();
+
+          // if we still have one fills with flipped bits, AND them here
+          if (ConciseSetUtils.isLiteral(w)) {
+            word &= w;
+          } else {
+            int flipBitLiteral = ConciseSetUtils.getLiteralFromOneSeqFlipBit(w);
+            if (flipBitLiteral != ConciseSetUtils.ALL_ONES_LITERAL) {
+              word &= flipBitLiteral;
```
Comment (drcrallen) [inline extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java]: This is new code and never touched by unit tests as far as I can measure.

File: extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java [hunk 4]
```diff
@@ -594,84 +624,56 @@ public int compare(WordHolder h1, WordHolder h2)
...
-        // advance the set with the current literal forward and push result back to priority queue
-        addAndCompact(retVal, word);
-        currIndex++;
-
-        if (itr.hasNext()) {
-          wordsToAdd.add(new WordHolder(itr.next(), itr));
+          itr.word = itr.next();
+          maxChangedIndex = Math.max(maxChangedIndex, 0);
         } else {
+          removeElement(theQ, 0, qSize);
```
Comment (drcrallen) [inline extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java]: This branch isn't touched by unit tests either

File: extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java [hunk 5]
```diff
@@ -594,84 +624,56 @@ public int compare(WordHolder h1, WordHolder h2)
...
-          WordHolder entry = theQ.poll();
-          int w = entry.getWord();
-          WordIterator i = entry.getIterator();
+          WordIterator i = theQ[qIndex];
+          int w = i.getWord();
 
           flipBitLiteral = ConciseSetUtils.getLiteralFromOneSeqFlipBit(w);
           if (flipBitLiteral != ConciseSetUtils.ALL_ONES_LITERAL) {
-            wordsToAdd.add(new WordHolder(flipBitLiteral, i));
+            i.word = flipBitLiteral;
```
Comment (drcrallen) [inline extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java]: Not touched by unit tests

File: extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java [hunk 6]
```diff
@@ -594,84 +624,56 @@ public int compare(WordHolder h1, WordHolder h2)
...
+            maxChangedIndex = qIndex;
+            qIndex++;
           } else if (i.hasNext()) {
-            wordsToAdd.add(new WordHolder(i.next(), i));
+            i.word = i.next();
+            maxChangedIndex = qIndex;
+            qIndex++;
           } else {
+            removeElement(theQ, qIndex, qSize);
```
Comment (drcrallen) [inline extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java]: Not touched by unit tests

File: extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java [hunk 7]
```diff
@@ -594,84 +624,56 @@ public int compare(WordHolder h1, WordHolder h2)
...
         if (flipBitLiteral != ConciseSetUtils.ALL_ONES_LITERAL) {
-          wordsToAdd.add(new WordHolder(flipBitLiteral, itr));
+          itr.word = flipBitLiteral;
+          maxChangedIndex = Math.max(maxChangedIndex, 0);
         } else if (itr.hasNext()) {
-          wordsToAdd.add(new WordHolder(itr.next(), itr));
+          itr.word = itr.next();
+          maxChangedIndex = Math.max(maxChangedIndex, 0);
         } else {
+          removeElement(theQ, 0, qSize);
```
Comment (drcrallen) [inline extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java]: Not touched by unit tests

**Comment by drcrallen** [comment]
Given the complexity and core importance of the code here, I called out a few places that were touched in the PR but don't have unit test coverage

**Comment by drcrallen** [comment]
I'll work through the logic in those places as soon as I can if noon else gets to them first.

File: extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java (hunk 3 above)
Comment (leventov) [inline extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java]: Now touched by `testIntersectionLiteralAndOneFill()`

File: extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java (hunk 4 above)
Comment (leventov) [inline extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java]: Now touched by `testIntersectionZeroSequenceRemovedFromQueue()`

File: extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java (hunk 5 above)
Comment (leventov) [inline extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java]: Now touched by `testIntersectionOneFillAndOneFillWithFlipBit()`

File: extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java (hunk 6 above)
Comment (leventov) [inline extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java]: Now touched by `testIntersectionSecondOneFillRemovedFromQueue()`

File: extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java (hunk 7 above)
Comment (leventov) [inline extendedset/src/main/java/io/druid/extendedset/intset/ImmutableConciseSet.java]: Now touched by `testIntersectionFirstOneFillRemovedFromQueue()`

**Comment by drcrallen** [comment]
Cool. I didn't dig deep into the logic, but if the unit tests cover it then I trust the tests.

**Pull request merged**